### PR TITLE
Fixed `from_dict` in app commands

### PR DESCRIPTION
**Pull request by EQUENOS**
## Summary

Better implementation of `from_dict` in
- `SlashCommand`
- `UserCommand`
- `MessageCommand`
- `Option`
- `OptionChoice`

Previous implementation wasn't solid enough to handle any sudden payload changes.

## Checklist

- [x] If code changes were made then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [ ] I have formatted the code properly by running `black .`
- [x] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by thesadru** [comment]
I don't think `_get_and_cast` is properly annotated here

File: disnake/app_commands.py [hunk 1]
```diff
@@ -102,6 +102,10 @@ def __eq__(self, other) -> bool:
     def to_dict(self) -> Dict[str, ChoiceValue]:
         return {"name": self.name, "value": self.value}
 
+    @classmethod
+    def from_dict(cls, data):
+        return OptionChoice(name=data.get("name"), value=data.get("value"))
```
Comment (shiftinv) [inline disnake/app_commands.py:107]:
```suggestion
        return OptionChoice(name=data["name"], value=data["value"])
```
These aren't optional and should always exist

File: disnake/app_commands.py [hunk 2]
```diff
@@ -346,8 +357,17 @@ def __repr__(self) -> str:
...
+        if data.pop("type", 0) != ApplicationCommandType.user.value:
+            raise ValueError("Invalid payload")
+
+        return UserCommand(
+            name=data["name"],
+            default_permission=data.get("default_permission", True),
+            id=_get_as_snowflake(data, "id"),
+            application_id=_get_as_snowflake(data, "application_id"),
+            guild_id=_get_as_snowflake(data, "guild_id"),
+            version=_get_as_snowflake(data, "version"),
```
Comment (shiftinv) [inline disnake/app_commands.py:369]:
```suggestion
            id=data.get("id"),
            application_id=data.get("application_id"),
            guild_id=data.get("guild_id"),
            version=data.get("version"),
```
`ApplicationCommand.__init__` already converts these, a `.get` should suffice
(same thing further down)

File: disnake/app_commands.py [hunk 3]
```diff
@@ -346,8 +357,17 @@ def __repr__(self) -> str:
 
     @classmethod
     def from_dict(cls, data: Dict[str, Any]):
-        if data.pop("type", 0) == ApplicationCommandType.user.value:
-            return UserCommand(**data)
+        if data.pop("type", 0) != ApplicationCommandType.user.value:
+            raise ValueError("Invalid payload")
```
Comment (shiftinv) [inline disnake/app_commands.py:361]: This error message should include the type
(same thing further down)

**Review by shiftinv — CHANGES_REQUESTED** [review]

**Comment by EQUENOS** [comment]
Corrections have been made

**Review by shiftinv — APPROVED** [review]

**Pull request merged**